tests/python/pants_test/engine/test_rules.py

```diff
@@ -677,41 +674,53 @@ def b_from_d(d):
                                     subgraph)
 
   def test_validate_yield_statements(self):
-    expected_rx_str = re.escape("""\
-Yield(value=Call(func=Name(id='A', ctx=Load()), args=[], keywords=[]{}))
-""".format('' if PY3 else ', starargs=None, kwargs=None'))
-    with self.assertRaisesRegexp(_RuleVisitor.YieldVisitError, expected_rx_str):
+    with self.assertRaisesRegexp(_RuleVisitor.YieldVisitError, re.escape('yield A()')):
```

Comment:
It would be exceedingly easy! I think @stuhood didn't have an issue with making this not error, but just didn't want to require the final return (and I was being lazy in the last PR because I wanted to get this one in, and forgot to add this capability back). Will do.

Reply:
@illicitonion : I'm not a fan of (optional) explicit return, and definitely not a fan of required explicit return.

Given the restriction that is applied here (that the yield is always a trailing statement in a block), I don't think accidental early return is likely. And there is nothing illegal about a coroutine that is "partially consumed". While the semantics of "if I yield a Get, I should expect to resume, otherwise, I can expect to exit" doesn't take much time to pick up... and the check that Danny has added helps to ensure that you don't get those swapped.

Reply:
(I also agree with the above, with the viewpoint that since we are *not* writing normal python code, I actually *don't* want it to look like valid python, and I feel ok about that because of the checking here.)

Reply:
To be clear: the python is 100% valid and normal.

If I write a coroutine that generates an infinite stream of fibonacci numbers, and then only consume part of it, that's no less a valid coroutine or coroutine consumer than something that consumes a coroutine until it is given the result value it wants.

Reply:
What I meant was that -- the `yield A()` part is what returns control to the engine, not the `return` below it, so allowing a `return` below it that isn't actually involved in any control flow feels very confusing to me.

Reply:
Oh -- maybe that's what you meant. My python experience is showing....

Reply:
@illicitonion I'm interpreting this conversation as "not allowing yield then return" for now, but it is extremely easy to add that capability afterwards (it's always easier to restrict then relax than the other way around, or something), as demonstrated by the previous iteration of the previous PR #7019 in which that was enforced.